File: android/app/src/main/java/com/certisgroup/cathy/reactutils/modules/crypto/DigestModule.java
/*
 * Copyright (c) 2020 Certis CISCO Security Pte Ltd
 * All rights reserved.
 *
 * This software is the confidential and proprietary information of
 * Certis CISCO Security Pte Ltd. ("Confidential Information").
 * You shall not disclose such Confidential Information and shall use
 * it only in accordance with the terms of the license agreement you
 * entered into with Certis CISCO Security Pte Ltd.
 */
package com.certisgroup.cathy.reactutils.modules.crypto;

import android.util.Base64;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * RN module for digest algorithms
 *
 * @author Lingqi
 */
public class DigestModule extends ReactContextBaseJavaModule {

    public DigestModule(@NonNull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "CryptoDigest";
    }

    @ReactMethod
    public void digest(String algo, String data, Promise promise) {
        try {
            byte[] dataBytes = Base64.decode(data, Base64.DEFAULT);
            MessageDigest messageDigest = MessageDigest.getInstance(algo);
            byte[] digestBytes = messageDigest.digest(dataBytes);
            String digest = Base64.encodeToString(digestBytes, Base64.NO_WRAP);
            promise.resolve(digest);
        } catch (NoSuchAlgorithmException | IllegalArgumentException e) {
            promise.reject("DIGEST_ERROR", e.getMessage(), e);
        }
    }
}

File: ios/cathy/RCTUtils/Modules/Crypto/DigestModule.swift
/*
* Copyright (c) 2020 Certis CISCO Security Pte Ltd
* All rights reserved.
*
* This software is the confidential and proprietary information of
* Certis CISCO Security Pte Ltd. ("Confidential Information").
* You shall not disclose such Confidential Information and shall use
* it only in accordance with the terms of the license agreement you
* entered into with Certis CISCO Security Pte Ltd.
*/
import Foundation
import CommonCrypto

/**
 * RN module for digest algorithms
 *
 * @author Lingqi
 */
@objc(CryptoDigest)
class DigestModule: NSObject {
    
    @objc(digest::::)
    func digest(algo: String, data: String,
                resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        guard let dataBuffer = Data(base64Encoded: data) else {
            reject("DIGEST_ERROR", "Data is not base64 encoded", nil)
            return
        }
        let dataBytes = [UInt8](dataBuffer)
        var digest: [UInt8]
        switch algo {
            case "SHA-1":
                digest = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
                CC_SHA1(dataBytes, CC_LONG(dataBytes.count), &digest)
            case "SHA-256":
                digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
                CC_SHA256(dataBytes, CC_LONG(dataBytes.count), &digest)
            case "SHA-384":
                digest = [UInt8](repeating: 0, count: Int(CC_SHA384_DIGEST_LENGTH))
                CC_SHA384(dataBytes, CC_LONG(dataBytes.count), &digest)
            case "SHA-512":
                digest = [UInt8](repeating: 0, count: Int(CC_SHA512_DIGEST_LENGTH))
                CC_SHA512(dataBytes, CC_LONG(dataBytes.count), &digest)
            default:
                reject("DIGEST_ERROR", "algorithm not supported", nil)
                return
        }
        let digestData = Data(bytes: digest, count: digest.count)
        let digestString = digestData.base64EncodedString()
        resolve(digestString)
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}
